### PR TITLE
core: Add ofi_sizeofaddr(), OFI_UNUSED(), atomic memory access macro, initial stuff of prov/netdir

### PR DESCRIPTION
**Pull request by gladkovdmitry17**
1.  Add ofi_sizeofaddr() function
2.  Add OFI_UNUSED() macro. Fix build warning
3.  Add atomic memory access macro
4.  Add initial stuff for new OFI/NetworkDirect provider

### CODE REVIEW

**Comment by gladkovdmitry17** [comment]
@shefty @a-ilango, kindly ask you to look into this PR, this PR contains changes outside prov/netdir from PR #2853 
The changes are only related to prov/netdir contains in PR #2853.

**Comment by a-ilango** [comment]
👍

File: include/unix/osd.h [hunk 1]
```diff
@@ -51,6 +51,43 @@
...
+
+#define ofi_val_compare_and_swap32(ptr, oldval, newval)		\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
+#define ofi_val_compare_and_swap64(ptr, oldval, newval)		\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
+#define ofi_val_compare_and_swap128(ptr, oldval, newval, compare)	\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
```
Comment (shefty) [inline include/unix/osd.h]: Please convert all of these into static inline calls, so that we have proper type checking and know what type of value each function returns.  We need to verify that all implementations return the same value.  For example always return the value after an update, or always return the value prior to the update.

File: include/unix/osd.h [hunk 2]
```diff
@@ -51,6 +51,43 @@
...
+	__sync_val_compare_and_swap((ptr), (ptr1), (ptr2))
+
+#define ofi_val_compare_and_swap32(ptr, oldval, newval)		\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
+#define ofi_val_compare_and_swap64(ptr, oldval, newval)		\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
+
+#define ofi_val_compare_and_swap128(ptr, oldval, newval, compare)	\
+	__sync_val_compare_and_swap((ptr), (oldval), (newval))
```
Comment (shefty) [inline include/unix/osd.h]: The behavior between unix and windows is different, or there's an error here.  The compare parameter is not used in unix, and I'm not sure why the 128-bit version requires 4 parameters, but the others only 3.

File: src/fabric.c [hunk 3]
```diff
@@ -437,6 +437,7 @@ void fi_ini(void)
 	ofi_register_provider(VERBS_INIT, NULL);
 	ofi_register_provider(GNI_INIT, NULL);
 	ofi_register_provider(BGQ_INIT, NULL);
+	ofi_register_provider(NETDIR_INIT, NULL);
```
Comment (shefty) [inline src/fabric.c]: Doesn't adding this here break the windows build?  There is no fi_netdir_ini() function included as part of the source yet.

File: include/unix/osd.h (hunk 2 above)
Comment (gladkovdmitry17) [inline include/unix/osd.h]: The 4th parameter is needed just to align Unix and Windows macro (or static inline function in the future)

File: include/unix/osd.h (hunk 1 above)
Comment (gladkovdmitry17) [inline include/unix/osd.h]: Ok, good point. Thank you.

File: src/fabric.c (hunk 3 above)
Comment (gladkovdmitry17) [inline src/fabric.c]: Yes, you're right. This build is broken w/o my 2nd PR
But I guess that we're going to merge 2nd PR (#2853) and only after that this one will be merged
Do you agree? or Do I need to move this one into 2nd PR?

**Comment by shefty** [comment]
> +	ofi_register_provider(NETDIR_INIT, NULL);
> 
> Yes, you're right. This build is broken w/o my 2nd PR
> But I guess that we're going to merge 2nd PR (#2853
> <https://github.com/ofiwg/libfabric/pull/2853> ) and only after that
> this one will be merged
> Do you agree? or Do I need to move this one into 2nd PR?

Please move this change into the 2nd PR or after it.  Either works for me.  Thanks

**Comment by gladkovdmitry17** [comment]
The registration of Network Direct provider has been moved to 2nd PR in the new commit. Atomic memory access macro were replaced by static inline functions

File: include/unix/osd.h [hunk 4]
```diff
@@ -45,6 +45,87 @@
...
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
+
+static inline
+unsigned char ofi_val_compare_and_swap128(volatile long long *ptr, long long oldval,
+					  long long newval, long long *compare)
+{
+	OFI_UNUSED(compare);
+	return __sync_val_compare_and_swap((ptr), (oldval), (newval));
+}
```
Comment (shefty) [inline include/unix/osd.h]: This function is not actually handling 128-bit values.

File: include/windows/osd.h [hunk 5]
```diff
@@ -102,7 +102,81 @@ do						\
...
+{
+	return InterlockedCompareExchange64(ptr, oldval, newval);
+}
+
+static inline
+unsigned char ofi_val_compare_and_swap128(volatile long long *ptr, long long oldval,
+					  long long newval, long long *compare)
+{
+	return InterlockedCompareExchange128(ptr, oldval, newval, compare);
+}
```
Comment (shefty) [inline include/windows/osd.h]: Can we just drop the 128-bit compare/swap functions?  The prototype for windows is:
```
unsigned char __cdecl InterlockedCompareExchange128(
  _Inout_ LONGLONG volatile *Destination,
  _In_    LONGLONG          ExchangeHigh,
  _In_    LONGLONG          ExchangeLow,
  _Inout_ LONGLONG          *ComparandResult
);
```
So the parameter names are misleading.  oldval and newval are both part of the new value.  The windows API itself is kind of goofy in that the new value is passed in as parts, but the old value is passed in as a pointer to an array.  Rather than try to come up with a sane solution, if we can drop it and defer adding it, that would be best (unless it's used later).

Comment (gladkovdmitry17) [inline include/windows/osd.h]: Yes, I agree. When I'll move netdir_buf.h content as a part of common code, I should found the proper solution for that since this function is used.

File: include/unix/osd.h (hunk 4 above)
Comment (gladkovdmitry17) [inline include/unix/osd.h]: it was improper idea, but the idea is to use `__int128` for this. Seems, it was a bit incorrectly to rely on that since it may be a common code for Win OS and Unix-like OS

**Comment by shefty** [comment]
FYI - The first two commits of this series have been merged in PR #2891 
Add ofi_sizeofaddr() function
Add OFI_UNUSED() macro. Fix build warning

File: include/unix/osd.h [hunk 6]
```diff
@@ -51,6 +51,73 @@
...
+long ofi_val_compare_and_swap32(volatile long *ptr, long oldval, long newval)
+{
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
+
+static inline
+long long ofi_val_compare_and_swap64(volatile long long *ptr, long oldval, long newval)
+{
+	return __sync_val_compare_and_swap(ptr, oldval, newval);
+}
```
Comment (shefty) [inline include/unix/osd.h]: There are atomic functions defined in include/fi_atom.h.  Those should be used, modified, or added to as needed.  That code handles situations where native atomic support is not available.  We may need to restructure the code there to handle windows.  It may be best to submit this patch in its own PR, so we can cross platform check it.

Comment (gladkovdmitry17) [inline include/unix/osd.h]: I apologize that I can got it wrong. Should I just drop the changes for unix for now? or for Windows also? Seems these builtin atomics should be more efficient than atomics in fi_atom.h.
If I got you correctly, we will add these atomic functions, but a bit later. Please, correct me if I'm wrong.

**Comment by shefty** [comment]
If the atomic functions aren't needed immediately, I would drop them for now.  The atomics in fi_atom.h convert to optimized, built-in functions when they are available.
 
I don't know if Windows supports stdatomic.h, but it's a C11 standard.

**Comment by shefty** [comment]
I looked at the atomic use in your other PR.  Most of the calls can convert to using atomic_inc/dec from fi_atom.h.  There's one 64-bit atomic increment used with msg_cnt and seq.  If that can use a 32-bit value instead, we can just convert to using the existing calls.  Otherwise, we can add 64-bit atomics by extending what's there. E.g. create an atomic64_t, and add atomic64_inc/dec calls.  The code only needs increment, but I can add the other definitions.

**Comment by gladkovdmitry17** [comment]
@shefty Thank you for your comments. Yes, I agree and you my find updated atomic API function in new PR #2894

**Comment by gladkovdmitry17** [comment]
@shefty Maybe will it be better to move these changes into NetwrokDirect related PR and just close this PR due to no need?

**Comment by shefty** [comment]
Sounds good
